ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/processor/BlockReplay.java

```diff
@@ -41,13 +44,88 @@ public BlockReplay(
+                                  mutableWorldState,
+                                  transactionProcessor))
+                      .collect(Collectors.toList());
+              return Optional.of(new BlockTrace(transactionTraces));
+            })
+        .orElse(null);
+  }
+
+  public BlockTrace block(final Hash blockHash, final Action<TransactionTrace> action) {
+    return getBlock(blockHash).map(block -> block(block, action)).orElse(null);
+  }
+
   public <T> Optional<T> beforeTransactionInBlock(
       final Hash blockHash, final Hash transactionHash, final Action<T> action) {
-    final BlockHeader header = blockchain.getBlockHeader(blockHash).orElse(null);
+    return performActionPerBlockTransaction(
+        blockHash,
+        (body, header, blockchain, mutableWorldState, transactionProcessor) -> {
+          final BlockHashLookup blockHashLookup = new BlockHashLookup(header, blockchain);
+          for (final Transaction transaction : body.getTransactions()) {
+            if (transaction.hash().equals(transactionHash)) {
+              return Optional.of(
+                  action.performAction(
+                      transaction, header, blockchain, mutableWorldState, transactionProcessor));
+            } else {
+              final ProtocolSpec<?> spec = protocolSchedule.getByBlockNumber(header.getNumber());
+              transactionProcessor.processTransaction(
+                  blockchain,
+                  mutableWorldState.updater(),
+                  header,
+                  transaction,
+                  spec.getMiningBeneficiaryCalculator().calculateBeneficiary(header),
+                  blockHashLookup);
+            }
+          }
+          return Optional.empty();
+        });
+  }
+
+  public <T> Optional<T> afterTransactionInBlock(
+      final Hash blockHash, final Hash transactionHash, final Action<T> action) {
+    return beforeTransactionInBlock(
+        blockHash,
+        transactionHash,
+        (transaction, blockHeader, blockchain, worldState, transactionProcessor) -> {
+          final ProtocolSpec<?> spec = protocolSchedule.getByBlockNumber(blockHeader.getNumber());
+          transactionProcessor.processTransaction(
+              blockchain,
+              worldState.updater(),
+              blockHeader,
+              transaction,
+              spec.getMiningBeneficiaryCalculator().calculateBeneficiary(blockHeader),
+              new BlockHashLookup(blockHeader, blockchain));
+          return action.performAction(
+              transaction, blockHeader, blockchain, worldState, transactionProcessor);
+        });
+  }
+
+  private <T> Optional<T> performActionPerBlockTransaction(
+      final Hash blockHash, final BlockTransactionAction<T> action) {
+    return getBlock(blockHash)
+        .map(block -> performActionPerBlockTransaction(block.getHeader(), block.getBody(), action))
+        .orElse(Optional.empty());
+  }
+
+  private <T> Optional<T> performActionPerBlockTransaction(
```

Comment:
This method name is a bit misleading.  It sounds like it's looping through calling `action` for each transaction in the block but actually appears to just call the action once. So maybe `performActionWithBlock` would be a better name?